ctms/crud.py

```diff
@@ -267,6 +268,86 @@ def get_contacts_by_any_id(
     return data
 
 
+def get_pending_acoustic_records(
+    db: Session,
+    end_time: datetime,
+):
+    """Get all the pending records up to a certain date."""
+    pending_records: List[PendingAcousticRecord] = (
+        db.query(PendingAcousticRecord)
+        .filter(
+            PendingAcousticRecord.update_timestamp < end_time,
+        )
+        .order_by(asc(Email.update_timestamp), asc(Email.email_id))
+        .all()
+    )
+
+    return pending_records
+
+
+def get_pending_records_as_contacts(
+    pending_records: List[PendingAcousticRecord],
+) -> List[ContactSchema]:
+    return [
+        ContactSchema.parse_obj(
+            {
+                "amo": record.email.amo,
+                "email": record.email,
+                "fxa": record.email.fxa,
+                "mofo": record.email.mofo,
+                "newsletters": record.email.newsletters,
+                "vpn_waitlist": record.email.vpn_waitlist,
+            }
+        )
+        for record in pending_records
+    ]
+
+
+def create_pending_acoustic_record(
```

Comment:
If I understand the design (and I'm not sure I do), there's at most one row per ``email_id``. That means if an ``email_id`` already appears in this table, the code should update the existing row. In that case, we don't need ``create_pending_acoustic_record``, since every operation is more like an upsert. ``update_pending_acoustic_record`` is closer, since it has the insert with an ``on_conflict_do_update`` clause. However, it seems overly generic for table-as-a-queue.

I think ``create_pending_acoustic_record`` and ``update_pending_acoustic_record`` need to be replaced with more specific operations:

* ``schedule_acoustic_record`` - called by the API when things change. Takes a DB session and an ``email_id``, and either creates or updates the ``PendingAcousticRecord``. I think existing rows get the retry count reset to 0 and the update timestamp set to now.
* ``retry_acoustic_record - called by the background function when Acoustic fails. Updates the retry count, leaves the rest alone.
* ``delete_acoustic_record - called by the background function when Acoustic succeeds. Similar to what you have, delete the record.

For the second two, I'd be tempted to take the ``update_timestamp`` as well, as a check that the API didn't update the row while the background job was processing it.  However, it may be better to lean on the database for row locking (such as ``SELECT FOR UPDATE`` in ``schedule_acoustic_record``).

I think this means the function interface is a DB session and and ``email_id``, and ``PendingAcousticRecordInSchema`` (and the other new schemas) are unnecessary.

I doubt I've understood the design however, since I don't have good notes from our last conversation on how this works, so I'm deferring to others.

Reply:
I think your comments here are spot on. I'm hoping to make the operations more impactful for the specific tasks. 

As for the table allowing at most one row per `email_id` that is up in the air; we can make that row non-unique and allow for any number. I couldn't recall the chosen approach and opted to do it as unique for the time being, this is reversible though.

Reply:
I think non-unique `email_ids` would make it more like a pub-sub queue and simplify a lot. Scheduling a sync is always adding a row, and the background task is either updating or deleting rows. That reduces the need for coordination and locking bugs.

Reply:
I thought `there's at most one row per email_id` was not true. I feel like we're all on different pages for this now :slightly_frowning_face: 

I also need a bit to page this back into my head. Maybe we should chat this afternoon